src/main/java/org/zalando/nakadi/validation/SchemaCompatibilityChecker.java

```diff
@@ -0,0 +1,116 @@
+package org.zalando.nakadi.validation;
+
+import com.google.common.collect.Lists;
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.validation.schema.NotSchemaConstraint;
+import org.zalando.nakadi.validation.schema.PatternPropertiesConstraint;
+import org.zalando.nakadi.validation.schema.SchemaConstraint;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Stack;
+
+@Component
+public class SchemaCompatibilityChecker {
+
+    final private List<SchemaConstraint> CONSTRAINTS = Lists.newArrayList();
+
+    public SchemaCompatibilityChecker() {
+        CONSTRAINTS.add(new NotSchemaConstraint());
+        CONSTRAINTS.add(new PatternPropertiesConstraint("patternProperties"));
```

Comment:
Why not to inject these classes?
For example to make a list of SchemaConstraint a Spring bean and to inject it in constructor.
The way you made it makes  SchemaCompatibilityChecker very bound with NotSchemaConstraint and PatternPropertiesConstraint. For example in your unit test for SchemaCompatibilityChecker you actually also test these two classes.